crates/services/p2p/src/request_response/messages.rs

```diff
@@ -34,33 +29,32 @@ pub(crate) const MAX_REQUEST_SIZE: usize = core::mem::size_of::<RequestMessage>(
 // This `OutboundResponse` gets prepared to be sent over the wire in `NetworkResponse` format.
 // The Peer that requested the message receives the response over the wire in `NetworkResponse` format.
 // It then unpacks it into `ResponseMessage`.
-// `ResponseChannelItem` is used to forward the data within `ResponseMessage` to the receving channel.
+// `ResponseChannelItem` is used to forward the data within `ResponseMessage` to the receiving channel.
 // Client Peer: `RequestMessage` (send request)
 // Server Peer: `RequestMessage` (receive request) -> `OutboundResponse` -> `NetworkResponse` (send response)
 // Client Peer: `NetworkResponse` (receive response) -> `ResponseMessage(data)` -> `ResponseChannelItem(channel, data)` (handle response)
 
-#[serde_as]
 #[derive(Serialize, Deserialize, Eq, PartialEq, Debug, Clone)]
 pub enum RequestMessage {
     Block(BlockHeight),
     SealedHeaders(Range<u32>),
-    Transactions(#[serde_as(as = "FromInto<[u8; 32]>")] BlockId),
+    Transactions(Range<u32>),
```

Comment:
The block itself can be very huge while being invalid at the same time because the header is invalid(the block producer or consensus did not sign it).

Requesting headers first ensures that at least this header(and a related block to it) was created according to the rules. It helps to identify potential forks and prevent DoS vectors. Punishment for sending the wrong header should be less than sending an invalid transaction related to this header because, in the second case, it is a malicious move, while invalid headers are not so problematic or may be related to the fork or misconfiguration.

Reply:
I think it's a fair question. During the import, we require a few checks for data integrity. Some of those include:
- The signature of a sealed block header must be valid
- A Header must have the expected height (as per the requested range)
- The Merkle root formed by the downloaded transactions must match the root in the header
If any of these checks fail for a given header, then the block is invalid, and we don't want to execute it. A failure could mean that the data is corrupt or the peer is acting maliciously. 

If we didn't need any of these checks, it is conceivable that we could execute and commit a block immediately. In that case, we would just download the block for execution. But these checks mitigate the consequences of adversarial behaviour. 

Let me know if that answers your question satisfactorily.